[DataGridPro] Allow to automatically group the rows in a Tree Data (#2725)
mui-org/material-ui-x: docs/src/pages/components/data-grid/group-pivot/group-pivot.md

{{"demo": "pages/components/data-grid/group-pivot/SetRowExpansionTreeData.js", "bg": "inline", "defaultCodeOpen": false}}

### Gaps in the tree

If some entries are missing to build the full tree, the `DataGridPro` will automatically create rows to fill those gaps.

{{"demo": "pages/components/data-grid/group-pivot/TreeDataWithGap.js", "bg": "inline", "defaultCodeOpen": false}}

### Filtering

A node is included if one of the following criteria is met:

- at least one of its descendant is passing the filters
- it is passing the filters

By default, the filtering is applied to every depth of the tree.
You can limit the filtering to the top level rows with the `disableChildrenFiltering` prop.

{{"demo": "pages/components/data-grid/group-pivot/DisableChildrenFilteringTreeData.js", "bg": "inline", "defaultCodeOpen": false}}

### Sorting

By default, the sorting is applied to every depth of the tree.
You can limit the sorting to the top level rows with the `disableChildrenSorting` prop.

{{"demo": "pages/components/data-grid/group-pivot/DisableChildrenSortingTreeData.js", "bg": "inline", "defaultCodeOpen": false}}

> > If you are using `sortingMode="server"`, you need to always set the children of a row after this row.

Reviewer: A small detail. It's not clear what is "this row".

```suggestion
> If you are using `sortingMode="server"`, you need to always put the children of a row after its parent.
```

For instance:

```ts

> // The row A.A is just after its parent

Author: Fixed, it's much more readable that way

> // The row A.A is immediately after its parent

Reviewer: The code should be inside the warning.

const validRows = [{ path: ['A'] }, { path: ['A', 'A'] }, { path: ['B'] }];

// The row A.A is not immediately after its parent

> const invalidRows = [{ path: ['A'] }, { path: ['B'] }, { path: ['A', 'A'] }];

Reviewer: Add emojis to make it clear what is the wrong pattern. Two examples: https://mui.com/guides/minimizing-bundle-size/#option-1 and https://mui.com/customization/how-to-customize/#main-content

<p class="description">Use grouping, pivoting and more to analyse the data in depth.</p>

## Tree Data [<span class="pro"></span>](https://mui.com/store/items/material-ui-pro/)

Tree Data allows to display data with parent / child relationships.

> ### General behavior

Reviewer: I don't think we need this title.

  getTreeDataPath={(row) => row.path}
  rows={rows}
  columns={columns}
/>;

// With transformation
const rows: GridRowsProp = [
  { path: 'Sarah', jobTitle: 'CEO', id: 0 },
  { path: 'Sarah/Thomas', jobTitle: 'Head of Sales', id: 1 },
  { path: 'Sarah/Thomas/Robert', jobTitle: 'Sales Person', id: 2 },
  { path: 'Sarah/Thomas/Karen', jobTitle: 'Sales Person', id: 3 },
];

<DataGridPro
  treeData
  getTreeDataPath={(row) => row.path.split('/')}
  rows={rows}
  columns={columns}
/>;
```

{{"demo": "pages/components/data-grid/group-pivot/BasicTreeData.js", "bg": "inline", "defaultCodeOpen": false}}

### Custom grouping column

Use the `groupingColDef` prop to customize the rendering of the grouping column.

{{"demo": "pages/components/data-grid/group-pivot/CustomGroupingColumnTreeData.js", "bg": "inline", "defaultCodeOpen": false}}

### Group expansion

Use the `defaultGroupingExpansionDepth` prop to expand all the groups up to a given depth when loading the data.
If you want to expand the whole tree, set `defaultGroupingExpansionDepth = -1`

{{"demo": "pages/components/data-grid/group-pivot/DefaultGroupingExpansionDepthTreeData.js", "bg": "inline", "defaultCodeOpen": false}}

> Use the `unstable_setRowExpansion` method on `apiRef` to programmatically set the expansion of a row.

Reviewer: Is not weird to tell users to use an unstable method? About the naming, it may cause confusion when #211 is implemented.

Author: Yes it's probably not great to prefix it and document it.
As it is a very basic method that will probably never change, I suppose we can make it public.
I had in mind we would use the same expansion boolean and method for boths as discussed on Monday. But as we won't, I can rename it`setRowChildrenExpansion`. In which case I would rename the `node.expanded` into `node.childrenExpanded`

Author: Done

{{"demo": "pages/components/data-grid/group-pivot/SetRowExpansionTreeData.js", "bg": "inline", "defaultCodeOpen": false}}

### Gaps in the tree

If some entries are missing to build the full tree, the `DataGridPro` will automatically create rows to fill those gaps.

{{"demo": "pages/components/data-grid/group-pivot/TreeDataWithGap.js", "bg": "inline", "defaultCodeOpen": false}}

### Filtering

A node is included if one of the following criteria is met:

- at least one of its descendant is passing the filters
- it is passing the filters

By default, the filtering is applied to every depth of the tree.
You can limit the filtering to the top level rows with the `disableChildrenFiltering` prop.

{{"demo": "pages/components/data-grid/group-pivot/DisableChildrenFilteringTreeData.js", "bg": "inline", "defaultCodeOpen": false}}

### Sorting

By default, the sorting is applied to every depth of the tree.

> You can limit the filtering to the top level rows with the `disableChildrenSorting` prop.

Reviewer: ```suggestion
You can limit the sorting to the top level rows with the `disableChildrenSorting` prop.
```
